README.md

```diff
@@ -1,17 +1,73 @@
+* `prodution` branch: This is the branch for tracking production code. Every commit on the branch will be shipped to the production environment.
+
+## [Concourse pipeline to support "Environment branches with Gitlab flow"](ci/3-envs.yaml)
+
+![3-envs.yml](images/3-envs.png)
+
+[3-envs.yml](ci/3-envs.yml) supports that:
+* Semantic versioning by [semver-resource](https://github.com/concourse/semver-resource)
+* Continuous Integration(CI) for pull requests.
+* Continuous Deployment(CD) to `staging` environment of master branch.
+  * deployment to `staging` environment bumps development version (`dev-version` resource)
+    * development version will be of the form `x.y.z-dev.n`
+* CI/CD of `pre-production` branch
+  * `merge-master-to-pre-production` job is the trigger to start deployment to `pre-production` environment.
+  * Succeeding jobs performs CI/CD to `pre-production` environment.
+   * `version-pre-production` job always computes release candidate version from `dev-version` resource and stores it to `rc-version` resource.
+    * If development version was `x.y.z-dev.n`, release candidate version will be `x.y.z-rc.1`
+   * `start-next-dev` is executed after `tag-pre-production`.  This stores next development version to `dev-version` resource.
+     * If `x.y.z-rc.1` will be tagged, `start-next-dev` bumped patch version.  Next dev version will be `x.y.(z+1)-dev.1`
+     * This means release candidate version will never be bumped to `rc.2`.  Team needs to fix commits on mater branch and start over CI/CD flow from `merge-master-to-pre-production`
+     * Team can manually bump patch/minor/master branch when big changes happened on master branch.
+  * `e2e-test-on-pre-production` performs end-to-end test on `pre-production` environment.  This would work as the final gate to promote pre-production branch to production branch.
+* CI/CD of `production` branch
+  * When `e2e-test-on-pre-production` passed, CI/CD process of production branch will be triggered automatically.  `merge-pre-production-to-production` will be kicked.
+  * Succeeding jobs performs CI/CD to `production` environment.
+   * `version-production` job always computes final version from `rc-version` resource and stores it to `final-version` resource.
+    * If release candidate version was `x.y.z-rc.1`, final version will be `x.y.z`
+* tagging and actual shipment jobs are separated.  This enables you to ship the latest commit on production branch to production environment __without__ bumping versions.  This would be useful for deployment failure or network issues.
+
+Please note that this pipeline does NOT support hotfix-ing on master branch.  This means you need to commits those fixes to master branch and start over release process.  See [3-envs.yml](ci/3-envs.yml) for detailed.
```

Comment:
I guess you'd like `hotfix-ing on production branch` instead of `hotfix-ing on master branch` here?